<div style="width: 100%; clear: both;">
    <div style="float: left; width: 50%;">
       <img src="http://www.uoc.edu/portal/_resources/common/imatges/marca_UOC/UOC_Masterbrand.jpg", align="left">
    </div>
</div>

<div style="float: right; width: 50%;">
    <p style="margin: 0; padding-top: 22px; text-align:right;">M1.871 · Privacidad</p>
    <p style="margin: 0; text-align:right;">Máster Universitario en Ciberseguridad y Privacidad</p>
    <p style="margin: 0; text-align:right; padding-button: 100px;">Estudios de Informática, Multimedia y Telecomunicación</p>
</div>

</div>
<div style="width: 100%; clear: both;">
<div style="width:100%;">&nbsp;</div>

# Práctica 1

--- 

Para resolver esta práctica, tenéis que poner las soluciones en el mismo
*notebook*, en las celdas de código que están en blanco habilitadas para ello. Después de cada celda de código podéis poner una celda *markdown* en con la justificación de la respuesta dada en la solución. También 
podéis poner justificaciones en forma de comentario en las celdas de código. No es obligatorio, pero sí muy recomendable que justifiquéis las respuestas que dais, ya que eso facilita la corrección en caso de una respuesta incorrecta (pudiendo valorar el planteamiento).

En muchos casos, cuando se pide implementar una función, se proporciona el nombre de la función y los parámetros, o algún tipo de esqueleto, indicando como ayuda el tipo de dichos parámetros y el tipo que retorna la función en forma de *type hint*. Es necesario que respetéis el nombre de la función y en la medida de los posibles los tipos. Sin embargo podéis implementar el código de la función con total libertad. Podéis
definir otras funciones adicionales y no respetar del todo los *type hints* si así lo consideráis oportuno. 

Este notebook va acompañado de un cuestionario online con preguntas sobre los ejercicios. Este cuestionario debe ser contestado antes de la fecha de etrega de la práctica. **Se recomienda que afrontéis el cuestionario una vez finalizada la practica, habiendo realizado y entendido todos los ejercicios.** En muchos casos se proporcionan datos o posibles resultados con el objeto de ayudar a poder contestar el cuestionario aunque no se haya acabado o no se haya sabido implementar todos los ejercicios de este *notebook*.

En algunas preguntas hay algunas indicaciones finales como *Hints*, son sugerencias sobre el uso de funciones que os pueden ser utiles. En cualquier caso se trata simplemente de eso, sugerencias.

# Datos

Trabajaremos con datos que simulan el expediente academico de alumnos. Tenemos el fichero `data/students1.csv` que contiene 10000 registros con los siguientes atributos:
- `name`: nombre del alumno/a,
- `subject0`: nota de la asignatura 0 (valor real entre 0 y 10),
- `subject1`: nota de la asignatura 1,
- `subject2`: nota de la asignatura 2,
- `subject3`: nota de la asignatura 3,

No podemos publicar este fichero de microdatos, ya que se considera que la nota obtenida por el alumnado es privada. Es decir no podemos saber que nota ha obtenido una estudiante en concreto.



# Publicación de microdatos

Queremos ahora poder publicar una versión protegida del fichero de microdatos `data/students1.csv`. Para poder trabajar con estos datos utilizaremos la libreira `pandas` de Python. De esta manera podemos leer el fichero en memoria:

In [1]:
import numpy as np
import pandas as pd

df1 = pd.read_csv("data/students1.csv")
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      10000 non-null  object 
 1   subject0  10000 non-null  float64
 2   subject1  10000 non-null  float64
 3   subject2  10000 non-null  float64
 4   subject3  10000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 390.8+ KB


Como vemos, ahora tenemos en la variable `df1` la tabla de microdatos en una estructura [`DataFrame`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) de `pandas`. Notad que tamibén hemos importado la libreria `numpy`, ya que la utilizaremos más adelante.

## Ejercicio 1 
*[0.5p]*

Para publicar estos datos primero tenemos que ver si se pueden publicar todos los atributos.

En el caso de que haya algún atributo que no se pueda publicar, elimínalo del `DataFrame` `df1`.

*Hint*:
- [pandas.DataFrame.drop](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html)


In [2]:
## Puesto que hay que anonimizar a los alumnos, eliminaremos del DF la columna que los identifica
df1_anonim = df1.drop(columns=["name"])
## Comprobamos el resultado mostrando los primeros registros
df1_anonim.head()

subject0  subject1  subject2  subject3
0       8.5       6.1       1.0       8.8
1       0.1       4.9       1.8       5.2
2       2.5       7.0       4.1       9.7
3       6.5       6.4       6.1       9.0
4       0.8       9.7       2.3       5.3

# Ejercicio 2 
*[1.25p]*

Para el resto de atributos vamos a utilizar **ruido aditivo no correlacionado**, tal como se describe en los materiales. 

Para ello vamos a utilizar una distribución normal $N(\mu, \sigma^2)$ con $\mu = 0$, y como varianza tomaremos $\sigma^2 = p \sigma_o^2$
donde $\sigma_o^2$ es la varianza de la variable que queremos proteger.

Desarrollad una función `noise_add_normal(df, p)` que recibe como parámetro `df`, el DataFrame a anonimizar y el parámetro `p`, y retorna una copia del DataFrame protegido, aplicando ruido a cada atributo. 

Finalmente, obtened una versión protegida en la variable `df1_noise_1` utilizando $p=0.1$.

Algunas consideraciones:
- cada atributo (variable o columna) tendrá una distribución de ruido independiente.
- la función no puede modificar el DataFrame original, tiene que retornar un DataFrame nuevo.
- tened cuidado a la hora de pasar la varianza a la distribución normal de la librería `numpy`, fijaros que espera recibir la desviación estándar (no la varianza).
- el resultado después de aplicar ruido NO tiene que estar redondeado a un decimal (como sí lo estaban los valores originales), ni se tiene que aplicar ninguna corrección.


*Hint*:
- [`numpy.random.Generator.normal`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.normal.html#numpy.random.Generator.normal)
- [Random sampling, quick start](https://numpy.org/doc/stable/reference/random/index.html#random-quick-start)


In [3]:
def noise_add_normal(df: pd.DataFrame, p: float) -> pd.DataFrame:
    ## Creamos una copia para proteger el DF recibido como parámetro
    df_noise = df.copy()
    for col in df.columns:
        # Calculamos la varianza original de la columna
        var = df[col].var()
        # numpy espera recibir la desviación estándar (no la varianza), así que transformamos
        std = np.sqrt(p * var)  # σ = sqrt(p * σ₀²)
        # Generamos un vector de ruido con media (µ) 0 y la desviación std recién calculada
        # Este vector contiene un valor distinto de ruido por cada fila (alumno), con media tendiendo a cero, para evitar sesgo
        # La generación es independiente por columna, lo que permite preservar las correlaciones entre las asignaturas
        noise = np.random.normal(0, std, len(df))
        # Sumamos el ruido a la columna
        df_noise[col] = df[col] + noise
    return df_noise
    
## Obtención de una copia protegida en la variable df1_noise_1, utilizando p=0.1
df1_noise_1 = noise_add_normal(df1_anonim, p=0.1)
## Comprobamos una muestra del resultado
df1_noise_1.head()

subject0  subject1  subject2   subject3
0  10.169779  5.669147 -0.298879  10.196580
1   0.002356  4.370000  2.907191   5.170168
2   3.625568  6.005721  3.951136  10.164007
3   7.325760  6.458562  5.082112   8.957011
4   0.197744  8.599891  1.935313   5.425637

Como guía, los valores protegidos en `df1_noise_1` pueden ser algo parecido a lo que se muestra en la siguiente tabla.

|  | subject0 |	subject1 |	subject2 |	subject3 |
| --- | --- | --- | --- | --- |
| 0 | 6.626006 |	6.858254 |	3.024407 |	10.113438 |
| 1 | -0.190741 |	5.878539 |	1.641941 |	4.289627 |
| 2 | 0.326832 |	7.113423 |	2.489244 |	10.214147 |
| 3 | 7.815629 |	5.846698 |	7.173413 |	10.008515 |
| 4 | 1.246823 |	10.570642 |	2.149917 |	4.912380 |
| 5 | ... | ... | ... | ... |
| ... | ... | ... | ... | ... |

## Ejercicio 3
*[1.25p]*

En este ejercicio desarrollaremos dos medidas de **pérdida de información**.

Para ello vamos a utilizar la siguiente notación:
- $X$: conjunto de datos original. Lo podemos interpretar como una matriz de
tamaño $n \times d$ donde $n$ es el número de registros (filas) y $d$ el
número de variables (columnas).
- $X'$: conjunto de datos protegido, también $n \times d$.
- $R$: matriz de correlación de los datos originales $X$. Esta matriz es una matriz $d \times d$ donde el elemento $(i,j)$ es la correlación entre la variable $V_i$ y $V_j$. Como medida de correlación tomaremos el coeficiente de correlación de Pearson.
- $R'$: matriz de correlación de los datos protegidos $X'$.

En cada caso mediremos la media del error absoluto MAE (*mean absolute error*). Es decir:
- Sobre $X$, $X'$:
  - *MAE*: $\frac{1}{nd}\sum_{j=1}^d \sum_{i=1}^n \mid x_{ij} - x'_{ij} \mid$
- Sobre $R$, $R'$:
  - *MAE-corr*: $\frac{1}{d^2}\sum_{j=1}^d \sum_{i=1}^n \mid r_{ij} - r'_{ij} \mid$

Se pide la implementación de la función `information_loss(df_original, df_protected)` donde `df_original` es el DataFrame con los datos originales ($X$) y `df_protected` el DataFrame con los datos protegidos ($X'$). La función tiene que retornar dos valores, en este orden:
- valor del MAE sobre $X$ y $X'$
- valor del MAE sobre $R$ y $R'$

Una vez implementada, aplicad la función para medir la pérdida de información al proteger los datos con $p=0.1$ (`df1_noise_1` del ejercicio anterior).



In [4]:
from typing import Tuple

def information_loss(df_original: pd.DataFrame, df_protected: pd.DataFrame) -> Tuple[float, float]:
    ## Definimos las matrices de trabajo
    X = df_original     # conjunto de datos original
    X_ = df_protected   # conjunto de datos protegido

    ## Cálculo del MAE (Mean Absolute Error) sobre X y X'
    mae_X = np.abs(X - X_).to_numpy().mean()

    ## Cálculo del MAE sobre las matrices R y R' tomando el coeficiente de correlación de Pearson
    R = X.corr(method="pearson")    # matriz de correlación de los datos originales
    R_ = X_.corr(method="pearson")  # matriz de correlación de los datos protegidos
    mae_R = np.abs(R - R_).to_numpy().mean()

    return mae_X, mae_R

## Llamada a la función con los datos correspondientes
mae_datos, mae_correlaciones = information_loss(df1_anonim, df1_noise_1)

## Mostramos los resultados con la misma redacción que el enunciado
print("Valor del MAE sobre X y X':", mae_datos)
print("Valor del MAE sobre R y R':", mae_correlaciones)


Valor del MAE sobre X y X': 0.778351150395397
Valor del MAE sobre R y R': 0.0018078201449779317


<div style="color:blue"> 
<h4><strong>COMENTARIOS EJERCICIO 3</strong></h4>
En los resultados que hemos obtenido se observa que:
    
- el ruido aditivo aplicado altera en cierta medida los valores individuales de cada asignatura (MAE = 0.7813), al introducir una componente
aleatoria con varianza proporcional al parámetro p = 0.1,

- y, sin embargo, se mantienen casi intactas las correlaciones entre las 4 asignaturas (MAE = 0.0015), dado que el ruido tiene media cero
y se aplica de forma no correlacionada entre columnas.

Eso significa que la UTILIDAD estadística del juego de datos se mantiene prácticamente intacta.

## Ejercicio 4
*[2p]*

Queremos ahora producir una versión de los microdatos con el mayor nivel de ruido posible pero sin perder mucha información. Concretamente, queremos garantizar que el MAE (unicamente el MAE, no el MAE-corr) no supere el valor $3$.

Es decir, ¿qué nivel de ruido (parámetro $p$) podemos aplicar de manera que obtengamos el mayor nivel de privacidad pero manteniendo la perdida de información según el MAE por debajo de $3.0$?

Se pide que programéis el código para determinar dicho valor $p$. Para restringir la búsqueda, consideraremos que el parámetro $p$ puede tomar valores a partir de $0$ y siempre se expresará con un solo decimal. Es decir, podemos tener valores como $p=0.3$ o $p=1.2$ pero NO $p=0.31$ o $p=1.023$.

Tenéis que mostrar al final el valor de $p$ que buscamos y el MAE obtenido con dicho valor.


In [5]:
## Definimos el nivel máximo de ruido permitido mae_max
## e inicializamos las variables de cálculo de los valores óptimos mae_opt y p_opt
mae_max = 3.0
mae_opt = None
p_opt = None

## Número de repeticiones por valor de p
num_reps = 20

## Exploramos valores de p con 1 decimal, de 0.0 a 2.0 como primera aproximación
for p in [round(x * 0.1, 1) for x in range(0, 21)]:
    maes = []

    for _ in range(num_reps):
        # Generamos una versión protegida de X aplicando ruido con parámetro p
        X_ = noise_add_normal(df1_anonim, p)

        # Calculamos la pérdida de información (MAE entre X y X'), ignorando el MAE-corr ya que no es necesario
        mae, _ = information_loss(df1_anonim, X_)
        maes.append(mae)

    # Calculamos el percentil 95 del MAE para este valor de p
    mae_p95 = np.percentile(maes, 95)

    # Verificamos si cumple el umbral definido
    if mae_p95 < mae_max:
        mae_opt = mae_p95
        p_opt = p
    else:
        break  # Al superar el umbral, salimos del bucle

## Mostramos el mejor valor de p encontrado y su correspondiente MAE
print("Valor óptimo de p:", p_opt)
print("Percentil 95 del MAE asociado a ese valor de p:", mae_opt)

Valor óptimo de p: 1.4
Percentil 95 del MAE asociado a ese valor de p: 2.9409981121200848


<div style="color:blue"> 
<h4><strong>COMENTARIOS EJERCICIO 4</strong></h4>

En este ejercicio se buscaba optimizar el equilibrio entre Privacidad y Utilidad de la información en la publicación de microdatos.
El proceso es algo más complejo que calcular el MAE dentro de un bucle en el que p se va incrementando en 1 décima, ya que el MAE es una variable aleatoria, con cierta variabilidad en cada ejecución.

Eso significa que, tal y como me ha sucedido en la primera versión del código, ejecutándola varias veces he observado que, si bien casi siempre obtenía un valor de p óptimo = 1.4, en alguna ocasión ese valor ha sido 1.5, y si lo hubiese dado por bueno, ocurriría que en la mayoría de casos produciría un MAE > 3.0, es decir, por encima del valor máximo impuesto en el enunciado.

Para aumentar la fiabilidad del cálculo sin complicar excesivamente el código, he introducido un bucle externo que repite el cálculo 20 veces por cada p óptimo, y evalúa el percentil 95 del MAE. Esto garantiza que el valor seleccionado sea el mayor p posible que, con alta probabilidad, mantiene el MAE por debajo del umbral. Como era de esperar, ese valor no es 1.5 sino 1.4

## Ejercicio 5
*[1.25p]*

Con el valor $p$ obtenido en el ejercicio anterior, vamos a obtener un nuevo DataFrame protegido:
```Python
df_protected_final = noise_add_normal(df1, p)
```
(Si no has podido acabar el ejercicio anterior, utiliza el valor $p=1.8$, que no tiene por qué ser igual a la solución de dicho ejercicio)

Este será el fichero de microdatos publicado. Para ver que diferencias tenemos respecto al original se pide comparar los siguientes valores entre el fichero de notas original (`df1`) y el protegido (`df_protected_final`):
- media de cada asignatura por separado.
- media de todas las asignaturas.
- nota máxima y mínima de cada asignatura por separado
- nota máxima y mínima de todas las asignaturas.

Para ello, tenéis que implementar una función `compute_statistics` que reciba como parámetro un `DataFrame` con las notas y retorne un diccionario con los valores requeridos. Como ejemplo, si aplicamos la función a los datos originales, `df1`, debería retornar:
```python
{'avg_s0': 5.4178700000000015,
 'avg_s1': 5.462769999999979,
 'avg_s2': 5.408580000000002,
 'avg_s3': 5.327550000000016,
 'avg_total': 5.4041925,
 'max_s0': 10.0,
 'max_s1': 10.0,
 'max_s2': 10.0,
 'max_s3': 10.0,
 'max_total': 10.0,
 'min_s0': 0.0,
 'min_s1': 0.0,
 'min_s2': 0.0,
 'min_s3': 0.0,
 'min_total': 0.0}
```
De esta manera podemos comparar los valores sobre `df1` y `df_protected_final`
```python
stats = pd.DataFrame([compute_statistics(df1), compute_statistics(df_protected_final)])

```
En `stats` tenemos un `DataFrame` donde podemos ver los resultados.
`
En este ejercicio se pide:
- implementar la función `compute_statistics` tal como se pide,
- mostrar el resultado en el `DataFrame` stats como se indica anteriormente.

In [6]:
from typing import Dict

def compute_statistics(df: pd.DataFrame) -> Dict[str, float]:
    ## Calculamos las medias por asignatura
    avg_s0 = df["subject0"].mean()
    avg_s1 = df["subject1"].mean()
    avg_s2 = df["subject2"].mean()
    avg_s3 = df["subject3"].mean()

    ## Calculamos la media global
    avg_total = df.mean().mean()

    ## Calculamos los máximos por asignatura
    max_s0 = df["subject0"].max()
    max_s1 = df["subject1"].max()
    max_s2 = df["subject2"].max()
    max_s3 = df["subject3"].max()

    ## Calculamos el máximo global
    max_total = df.max().max()

    ## Calculamos los mínimos por asignatura
    min_s0 = df["subject0"].min()
    min_s1 = df["subject1"].min()
    min_s2 = df["subject2"].min()
    min_s3 = df["subject3"].min()

    ## Calculamos el mínimo global
    min_total = df.min().min()

    ## Devolvemos los resultados en un diccionario
    return {
        "avg_s0": avg_s0,
        "avg_s1": avg_s1,
        "avg_s2": avg_s2,
        "avg_s3": avg_s3,
        "avg_total": avg_total,
        "max_s0": max_s0,
        "max_s1": max_s1,
        "max_s2": max_s2,
        "max_s3": max_s3,
        "max_total": max_total,
        "min_s0": min_s0,
        "min_s1": min_s1,
        "min_s2": min_s2,
        "min_s3": min_s3,
        "min_total": min_total
    }

## Generamos el DataFrame protegido final usando el valor óptimo de p obtenido en el ejercicio anterior
df_protected_final = noise_add_normal(df1_anonim, p_opt)

## Aplicamos la función compute_statistics sobre el dataset original y el protegido
stats = pd.DataFrame([
    compute_statistics(df1_anonim),
    compute_statistics(df_protected_final)
])

## Reorganizamos el DataFrame para mostrar una fila por variable, y mostramos el resultado
stats = stats.transpose().reset_index()
stats.columns = ["Variable", "Valor original", "Valor protegido"]
print(f"Comparativa entre los valores originales y los protegidos con p = {p_opt}")
stats

Comparativa entre los valores originales y los protegidos con p = 1.4


Variable  Valor original  Valor protegido
0      avg_s0        5.417870         5.434573
1      avg_s1        5.462770         5.485534
2      avg_s2        5.408580         5.463850
3      avg_s3        5.327550         5.284905
4   avg_total        5.404193         5.417216
5      max_s0       10.000000        22.985398
6      max_s1       10.000000        22.088635
7      max_s2       10.000000        21.411951
8      max_s3       10.000000        22.778471
9   max_total       10.000000        22.985398
10     min_s0        0.000000       -12.018260
11     min_s1        0.000000       -13.164599
12     min_s2        0.000000       -11.705911
13     min_s3        0.000000       -10.939736
14  min_total        0.000000       -13.164599

<div style="color:blue"> 
<h4><strong>COMENTARIOS EJERCICIO 5</strong></h4>
Al comparar las estadísticas del DF original con las del DF protegido usando el factor p obtenido en el ejercicio anterior (1.4) hemos observado que, a efectos estadísticos, el conjunto de datos sigue siendo válido, puesto que los promedios de las notas apenas varían. En cambio, si se pretendiera hacer un análisis de casos individuales, que no es el caso en este ejercicio, entonces obtendríamos conclusiones muy distorsionadas, con notas máximas por encima del 20 y notas mínimas con valores negativos, debido al nivel de ruido aplicado en esta ocasión.

# Privacidad diferencial

Ahora vamos a ver el enfoque que propone la privacidad diferencial, y para ello implementaremos de forma simulada una base de datos como una clase de Python.


La siguiente clase implementa una sencilla base de datos `MyDB`. La base de datos lee el fichero CSV que contiene los datos con las notas de los estudiantes.

Como vemos, se trata de una base de datos muy sencilla y simplificada, que únicamente tiene sentido con fines docentes. La idea es que la clase `MyDB` simula una base de datos y sus métodos públicos son las consultas que pueden hacer los usuarios. Asumimos que un usuario no puede ver el contenido del objeto y simplemente puede acceder a `MyDB` mediante sus métodos.

La clase `MyDB` se instancia a partir de un fichero CSV que se guarda como un `DataFrame` de `pandas`. Esta clase implementa 1 método para obtener información sobre las estudiantes de la base de datos:
- `count_all_fail`: que retorna el número de estudiantes que han suspendido todas las asignaturas.

Recordamos que asumimos que la única manera que tiene cualquier usuario para acceder a los datos es llamar a esta función `count_all_fail`.


In [7]:

class MyDB:
    def __init__(self, filename: str) -> None:
        self.df = pd.read_csv(filename)

    def count_all_fail(self) -> int:
        return self.df[
            (self.df["subject0"] < 5)
            & (self.df["subject1"] < 5)
            & (self.df["subject2"] < 5)
            & (self.df["subject3"] < 5)
            ].shape[0]


Como ejemplo, a continuación vemos como se instanciaría `SimpleDB` con el fichero `data/students1.csv`

In [8]:
db1 = MyDB("data/students1.csv")

print(f"DB1 count_all_fail: {db1.count_all_fail()}")

DB1 count_all_fail: 420


## Ejercicio 6
*[0.5p]*

Vamos ahora a estudiar la privacidad que ofrece `MyDB` según la definición de privacidad diferencial. Con el objeto de poder hacer diversas pruebas tenemos dos ficheros CSV:
- `data/students1.csv`: con 10000 registros
- `data/students2.csv`: con 9999 registros

Podemos ver que `students2.csv` es igual a `students1.csv` con la diferencia de que le falta un registro, el correspondiente a *Susan Duffy*. En principio podemos asumir que un posible atacante sabe que este es el registro que falta (el correspondiente a esta estudiante no está en DB2 y sí en DB1), pero no sabe sus datos: notas obtenidas en las asignaturas.

In [9]:
db1 = MyDB("data/students1.csv")
db2 = MyDB("data/students2.csv")


Observando el resultado de hacer la consulta `count_all_fail` a `db1` y `db2`. ¿Qué información podemos obtener? (Muestra el resultado de las consultas y añade la respuesta y su justificación como comentario).


In [10]:
## Creamos las 2 bases de datos cargando los ficheros de la carpeta data mediante la clase MyDB
db1 = MyDB("data/students1.csv") ## Base de datos con los 10.000 registros originales
db2 = MyDB("data/students2.csv") ## Copia de db1 pero sin el registro de Susan Duffy

## Ejecutamos la consulta count_all_fail sobre ambas bases de datos
## Esta consulta cuenta el número de estudiantes que han suspendido todas las asignaturas
suspendidos_db1 = db1.count_all_fail()
suspendidos_db2 = db2.count_all_fail()

## Mostramos los resultados obtenidos en ambas bases de datos
print("Número de estudiantes con todo suspendido en DB1:", suspendidos_db1)
print("Número de estudiantes con todo suspendido en DB2:", suspendidos_db2)

Número de estudiantes con todo suspendido en DB1: 420
Número de estudiantes con todo suspendido en DB2: 420


<div style="color:blue"> 
<h4><strong>COMENTARIOS EJERCICIO 6</strong></h4>

Al ejecutar el código de la celda de arriba hemos observado que el resultado es el mismo en las 2 bases de datos.
Esto es coherente con el concepto de *privacidad diferencial*, ya que la ausencia o presencia de un registro cualquiera no debe ser perceptible en las consultas que se realicen.

En este caso concreto, la eliminación de Susan Duffy no ha afectado al resultado porque ella no se encontraba en el conjunto de alumnos con todas las asignaturas suspendidas. De haberse eliminado un estudiante perteneciente a un conjunto de anonimato más estrecho (alumnos con todo suspendido), entonces la consulta sí que hubiese diferido, en una unidad.

Por lo tanto, observamos que cuanto mayores sean las bases de datos y los conjuntos de anonimato, mayor será el nivel de privacidad diferencial que podremos garantizar en las consultas agregadas.


## Ejercicio 7
*[1.25p]*

Vamos a implementar una nueva base de datos que implemente la versión segura del método `count_all_fail` utilizando el mecanismo de Laplace. Para ello crearemos una nueva clase de base de datos a la que llamaremos `MyPrivateDB`.



Implementa, en la siguiente clase, el método `dp_count_all_fail` que retorne el número de alumnos y alumnas que han suspendido todas las asignaturas, pero esta vez la respuesta ha de estar protegida utilizando **el mecanismo de Laplace**.

Para hacerlo necesitaremos:
- Calcular o determinar la sensibilidad global del método.
- Generar ruido con una distribución de Laplace basándonos en su sensibilidad global y el parámetro *epsilon*.
- Retornar el valor con el ruido añadido.

Como vemos, el parámetro *epsilon* se fija en el constructor de la clase. Importante: de momento asumimos que únicamente se realizará una consulta.

*Hint*: 
- [random.Generator.laplace](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.laplace.html#numpy.random.Generator.laplace)

In [11]:
import numpy as np

class MyPrivateDB:
    def __init__(self, filename: str, epsilon: float) -> None:
        self.df = pd.read_csv(filename)
        self.epsilon = epsilon

    def dp_count_all_fail(self) -> int:
        ## Calculamos el número real de alumnos con todas las asignaturas suspendidas
        count = self.df[
            (self.df["subject0"] < 5) &
            (self.df["subject1"] < 5) &
            (self.df["subject2"] < 5) &
            (self.df["subject3"] < 5)
        ].shape[0]

        ## Sensibilidad global = 1 (añadir o eliminar un registro puede variar el resultado en 1 unidad)
        sensibilidad = 1.0

        rng = np.random.default_rng()  ## Inicializamos el generador de números aleatorios

        ## Generamos ruido de Laplace con media 0 y escala sensibilidad/epsilon
        noise = rng.laplace(0, sensibilidad / self.epsilon)

        return int(round(count + noise))  ## Sumamos el ruido al valor real y redondeamos

<div style="color:blue"> 
<h4><strong>COMENTARIOS EJERCICIO 7</strong></h4>

El mecanismo de Laplace consiste en añadir un valor aleatorio extraído de una distribución de Laplace al resultado real de la consulta, para proteger la confidencialidad de los datos individuales. La cantidad de ruido añadido depende de dos factores:

- La sensibilidad global de la consulta (en este caso, 1, ya que añadir o eliminar un único registro puede alterar el resultado en una unidad).

- El parámetro epsilon, que regula el equilibrio entre privacidad y precisión: cuanto menor sea epsilon, mayor es el nivel de privacidad, pero más ruido se añade al resultado.

## Ejercicio 8
*[0.5p]*

Podemos ver el resultado de hacer la consulta `dp_count_all_fail` con la nueva base de datos `MyPrivateDB` a los datos de `data/students1.csv` y `data/students2.csv`.

¿Podemos ahora obtener la misma información que en el Ejercicio 6? Muestra el resultado para `epsilon=1`, responde y justifica en forma de comentario.

In [12]:
## Creamos las dos bases de datos privadas (protegidas con Laplace y epsilon=1)
db1_priv = MyPrivateDB("data/students1.csv", epsilon=1)  ## Base de datos con todos los registros
db2_priv = MyPrivateDB("data/students2.csv", epsilon=1)  ## Base de datos sin Susan Duffy

## Ejecutamos la consulta dp_count_all_fail en ambas bases de datos
suspendidos_db1_priv = db1_priv.dp_count_all_fail()
suspendidos_db2_priv = db2_priv.dp_count_all_fail()

## Mostramos los resultados obtenidos
print("Número de estudiantes con todo suspendido en DB1 (privado):", suspendidos_db1_priv)
print("Número de estudiantes con todo suspendido en DB2 (privado):", suspendidos_db2_priv)


Número de estudiantes con todo suspendido en DB1 (privado): 419
Número de estudiantes con todo suspendido en DB2 (privado): 421


<div style="color:blue"> 
<h4><strong>COMENTARIOS EJERCICIO 8</strong></h4>

A través de los resultados obtenidos en el ejercicio, se observa que al aplicar el método dp_count_all_fail() con un valor de epsilon=1, el resultado ya no coincide exactamente con el obtenido en el ejercicio 6, pese a que las bases de datos db1 y db2 siguen difiriendo en un único registro.

Este resultado es coherente con el funcionamiento del mecanismo de Laplace, cuya finalidad es proteger la privacidad de los datos individuales.

Para ello, al resultado real de la consulta se le añade un valor aleatorio extraído de una distribución de Laplace. Esto garantiza que la presencia o ausencia de un único individuo no sea detectable a partir de los resultados publicados.

Cuanto menor es el valor de epsilon, mayor es el nivel de privacidad diferencial que conseguimos, pero también se incrementa el nivel de ruido añadido al resultado, afectando a su precisión. Se puede apreciar más claramente el efecto haciendo pruebas con valores de epsilon muy bajos y muy altos. Por ejemplo, con epsilon = 0.1, el resultado oscilaría aleatoriamente en un rango aproximado de +-40, y con epsilon = 10, el valor calculado se mantendría constante en los 420 alumnos.

## Ejercicio 9
*[2p]*

Hemos asumido al principio que el usuario o usuaria únicamente realizará una consulta (llamará a la función `dp_count_all_fail()` una vez).

Si ahora asumimos que se puede hacer más de una consulta, y sin modificar la implementación de nuestro método `dp_count_all_fail`, un atacante ¿podría llegar a obtener la misma información que proporcionaba el método `count_all_fail` pero utilizando `dp_count_all_fail`?

En este ejercicio se pide implementar el posible ataque que permita desanonimizar `dp_count_all_fail` (es decir, utilizando `MyPrivateDB`) y con la particularidad de que se puede repetir dicha consulta sobre una misma base de datos diversas veces. En este caso lo haremos para `epsilon=8`. Justifica el resultado.

In [28]:
## Code goes here
## Creamos la base de datos privada a partir de students1.csv, con epsilon=8
db1_priv = MyPrivateDB("data/students1.csv", epsilon=8)

## Definimos el número de repeticiones del ataque
num_reps = 20

## Lista para almacenar los resultados obtenidos en cada consulta
resultados = []

## Ejecutamos la consulta dp_count_all_fail() varias veces
for _ in range(num_reps):
    resultado = db1_priv.dp_count_all_fail()
    resultados.append(resultado)

## Calculamos el promedio de los resultados obtenidos
promedio_resultados = np.mean(resultados)

## Mostramos el resultado del ataque
print("Valor promedio obtenido tras el ataque por repetición:", promedio_resultados)
print("Valor real (sin protección) del número de estudiantes con todo suspendido:", 420)

Valor promedio obtenido tras el ataque por repetición: 419.95
Valor real (sin protección) del número de estudiantes con todo suspendido: 420


<div style="color:blue"> 
<h4><strong>COMENTARIOS EJERCICIO 9</strong></h4>

En este ejercicio hemos simulado un posible ataque por repetición de consultas sobre la función dp_count_all_fail() con epsilon=8, y hemos podido verificar la limitación de la privacidad diferencial que señalan los apuntes de la asignatura basándose en el teorema de composición secuencial, el cual concluye que "cada vez que se hace una consulta a la base de datos se produce una pérdida de privacidad acumulativa".

Al permitir múltiples consultas independientes, un atacante puede anular el efecto del ruido aleatorio simplemente repitiendo muchas veces la consulta y calculando la media de los resultados obtenidos.

Los resultados nos muestran que tras realizar 20 consultas y calcular el promedio, obtenemos un valor que converge hacia el número real de alumnos con todas las asignaturas suspendidas (420). En otras palabras, conseguimos dejar sin efecto el ruido introducido por el mecanismo de Laplace. 